optking/molsys.py

```diff
@@ -677,43 +679,460 @@ def q_show_forces(self, forces):
+                geom = F.geom
+                cart_offset = 3 * self.frag_1st_atom(iF)
+                intco_offset = self.frag_1st_intco(iF)
+
+                for iIntco, Intco in enumerate(F.intcos):
+                    dq2dx2[:] = 0
+                    Intco.Dq2Dx2(geom, dq2dx2)  # d^2(q_I)/ dx_i dx_j
+
+                    # Loop over Cartesian pairs in fragment
+                    for a in range(3*F.natom):
+                        for b in range(3*F.natom):
+                            Hxy[cart_offset+a, cart_offset+b] += \
+                                g_q[intco_offset+iIntco] * dq2dx2[a, b]
+
+            # TODO: dimer coordinates
+            if self._dimer_intcos:
+                raise NotImplementedError('transformations with dimer gradients')
+
+        return Hxy
+
+
+    def gradient_to_cartesians(self, g_q):
+        """converts the gradient from internal into Cartesian coordinates
+
+        Parameters
+        ----------
+        g_q : ndarray
+            internal coordinate gradient
+        Returns
+        -------
+        g_x : ndarray
+            Cartesian coordinate gradient
+        """
+        logger = logging.getLogger(__name__)
+        logger.info("Converting gradient from internals to Cartesians.\n")
+        B = self.Bmat()
+        g_x = np.dot(B.T, g_q)
+        return g_x
+
+
+    def test_Bmat(self):
+        """ Test the analytic B matrix (dq/dx) via finite differences.
+        The 5-point formula should be good to DISP_SIZE^4 - a few
+        unfortunates will be slightly worse.
+
+        Returns
+        -------
+        passes : boolean
+            Returns True or False, doesn't raise exceptions
+        """
+        logger = logging.getLogger(__name__)
+        Natom = self.natom
+        Nintco = self.num_intcos
+        DISP_SIZE = 0.01
+        MAX_ERROR = 50 * DISP_SIZE * DISP_SIZE * DISP_SIZE * DISP_SIZE
+
+        logger.info("\tTesting B-matrix numerically...")
+
+        B_analytic = self.Bmat()
+
+        if op.Params.print_lvl >= 3:
+            logger.debug("Analytic B matrix in au")
+            logger.debug(print_mat_string(B_analytic))
+
+        B_fd = np.zeros((Nintco, 3 * Natom))
+
+        self.update_dihedral_orientations()
+        self.fix_bend_axes()
+
+        geom_orig = self.geom  # to restore below
+        coord = self.geom  # returns a copy
+
+        for atom in range(Natom):
+            for xyz in range(3):
+                coord[atom, xyz] -= DISP_SIZE
+                self.geom = coord
+                q_m = self.q()
+
+                coord[atom, xyz] -= DISP_SIZE
+                self.geom = coord
+                q_m2 = self.q()
+
+                coord[atom, xyz] += 3 * DISP_SIZE
+                self.geom = coord
+                q_p = self.q()
+
+                coord[atom, xyz] += DISP_SIZE
+                self.geom = coord
+                q_p2 = self.q()
+
+                coord[atom, xyz] -= 2 * DISP_SIZE  # restore to original
```

Comment:
I wouldn't try to force the xyz loop into numpy to vectorize it but I would suggest replacing the last loop with:
```
B_fd[: , 3 * natom + xyz] = (q_m2 - 8 * q_m + 8 *p_q - q_p2) / 12.0 * DISP_SIZE
```

Reply:
Yes, good idea.  Incorporated.

Reply:
That testing code was written before you could do optmol.Bmat() so it needs updated.  You can see it still uses the raw IntcosMisc.Bmat() too.  Will streamline later.  Need to keep informative error messages though.  Not a helpful debugging tool if can't find error.